cpp/include/raft/matrix/select.cuh

```diff
@@ -0,0 +1,158 @@
+ * @param[out] out_idx
+ *   contiguous device array of outputs of size (k * batch_size);
+ *   the payload selected together with `out_val`.
+ * @param select_min
+ *   whether to select k smallest (true) or largest (false) keys.
+ * @param stream
+ * @param mr an optional memory resource to use across the calls (you can provide a large enough
+ *           memory pool here to avoid memory allocations within the call).
+ */
+template <typename T, typename IdxT>
+void select_k(const T* in_val,
+              const IdxT* in_idx,
+              size_t batch_size,
+              size_t len,
+              int k,
+              T* out_val,
+              IdxT* out_idx,
+              bool select_min,
+              rmm::cuda_stream_view stream,
+              rmm::mr::device_memory_resource* mr = nullptr)
+{
+  common::nvtx::range<common::nvtx::domain::raft> fun_scope(
+    "matrix::select_k(batch_size = %zu, len = %zu, k = %d)", batch_size, len, k);
+  // TODO (achirkin): investigate the trade-off for a wider variety of inputs.
+  const bool radix_faster = batch_size >= 64 && len >= 102400 && k >= 128;
+  if (k <= detail::select::warpsort::kMaxCapacity && !radix_faster) {
+    detail::select::warpsort::select_k<T, IdxT>(
+      in_val, in_idx, batch_size, len, k, out_val, out_idx, select_min, stream, mr);
+  } else {
+    detail::select::radix::select_k<T, IdxT, (sizeof(T) >= 4 ? 11 : 8), 512>(
+      in_val, in_idx, batch_size, len, k, out_val, out_idx, select_min, stream, mr);
+  }
+}
+
+/**
+ * Select k smallest or largest key/values from each row in the input data.
+ *
+ * If you think of the input data `in_val` as a row-major matrix with `len` columns and
+ * `batch_size` rows, then this function selects `k` smallest/largest values in each row and fills
+ * in the row-major matrix `out_val` of size (batch_size, k).
+ *
+ * @tparam T
+ *   the type of the keys (what is being compared).
+ * @tparam IdxT
+ *   the index type (what is being selected together with the keys).
+ *
+ * @param[in] in_val
+ *   inputs values [batch_size, len];
+ *   these are compared and selected.
+ * @param[in] in_idx
+ *   optional input payload [batch_size, len];
+ *   typically, these are indices of the corresponding `in_val`.
+ *   If `in_idx` is `std::nullopt`, a contiguous array `0...len-1` is implied.
+ * @param[out] out_val
+ *   output values [batch_size, k];
+ *   the k smallest/largest values from each row of the `in_val`.
+ * @param[out] out_idx
+ *   output payload (e.g. indices) [batch_size, len];
+ *   the payload selected together with `out_val`.
+ * @param select_min
+ *   whether to select k smallest (true) or largest (false) keys.
+ * @param stream
+ * @param mr an optional memory resource to use across the calls (you can provide a large enough
+ *           memory pool here to avoid memory allocations within the call).
+ */
+template <typename T, typename IdxT>
+void select_k(raft::device_matrix_view<const T, size_t, row_major> in_val,
```

Comment:
I'm not sure if changing the rmm resource into the `optional` is a good idea. It looks like the standard approach to get an rmm resource is to call `rmm::mr::get_current_device_resource()`, which returns a pointer (we also do that in raft all over the place). Also, placing a resource inside an `optional` would imply a copy and and automatic storage/lifetime, while all the rmm resources are passed in rmm primitives (and in mdarray) by pointers. Copying and then getting out the scope would cause all sorts of troubles (e.g. access by invalid pointers at destruction of an mdarray or a uvector).

Reply:
>  It looks like the standard approach to get an rmm resource is to call rmm::mr::get_current_device_resource()

`get_current_device_resource()` is intended to fetch the "global" RMM allocator which has been set on the current device. This is often going to be set once per process and should be treated as the "main" allocator, and the user should decide whether that allocator should be pooled or managed (or both). By accepting an allocator in addition to the default allocator, RAFT is essentially saying "You can set a pool or managed allocator (or both or just standard) as the default but this function can make use of an additional pooled allocator to use for internal workspaces". Does that make sense?

> which returns a pointer (we also do that in raft all over the place).

That is correct, the default allocator which has configured on the current process for the current device is used in rmm's device buffers/vectors and rmm's device vector is then used in the default device container policy for RAFT's mdarray implementation. But please note that RAFT should no longer be using a pointer to a memory resource directly anymore (and if you find any cases where it is still being used in RAFT, it should be converted to using `rmm::device_uvector` or `raft::device_mdarray` instead. 

> Also, placing a resource inside an optional would imply a copy and and automatic storage/lifetime, while all the rmm resources are passed in rmm primitives (and in mdarray) by pointers

What I'm proposing is a std::optional using smart pointers, as we are doing in other places across RAPIDS: `std::optional<std::shared_ptr<rmm::mr::device_memory_resource>>`. 


Though, I think this discussion may be a bit moot at this point because I didn't think we'd be able to wrap up https://github.com/rapidsai/raft/pull/1111 before this PR is merged but I think that's just about ready to go in. The general idea here is to enable a resource container (like the `raft::handle_t`) agnostic of the resources it's containing. For example, this enables the handle to contain both host and device resources and ultimately be compiled in both host- and device-specific environments (such as Triton). For the purposes of this PR and the ANN APIs, it pulls `rmm` out of the publicly exposed APIs and centralize the configuration and construction of the underlying workspace allocator without having to deal with issues like we are discussing here.